Nullptr crash in WebCore::MathMLElement::isKeyboardFocusable

pratiksha-99 opened the pull request:
#### 5a55a82ed2271d86c304a5c96fce208c3a072624
<pre>
Nullptr crash in WebCore::MathMLElement::isKeyboardFocusable
<a href="https://bugs.webkit.org/show_bug.cgi?id=288296">https://bugs.webkit.org/show_bug.cgi?id=288296</a>
<a href="https://rdar.apple.com/144405709">rdar://144405709</a>

Reviewed by Anne van Kesteren and Ryosuke Niwa.

Added a nullptr check for the LocalFrame before focused keyboard events.

* LayoutTests/mathml/mathml-invalid-frame-crash-expected.txt: Added.
* LayoutTests/mathml/mathml-invalid-frame-crash.html: Added.
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::isKeyboardFocusable const):

Canonical link: <a href="https://commits.webkit.org/291060@main">https://commits.webkit.org/291060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9429c9a59720ea92544e180ea5b5ffaf2d64123e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91813 "Passed style check") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11347 "Hash 9429c9a5 for PR 41150 does not build (failure)") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/895 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96768 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42450 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93863 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19827 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70481 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27976 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94814 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8964 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83186 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50809 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/778 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41655 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79014 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/779 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98785 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18959 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/13997 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79507 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19210 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79030 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78730 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23261 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/581 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/12016 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14577 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18942 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24165 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18643 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22100 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20399 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,26 @@
+  var dialogElement = document.getElementById("htmlvar00011");
+
+  var newDocument = window.document.implementation.createHTMLDocument();
+  var docToUse = newDocument || window.document;
```
Isn't this always going to be `newDocument`?

annevk commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -268,8 +268,12 @@ bool MathMLElement::isKeyboardFocusable(KeyboardEvent* event) const
 
+    RefPtr frame = document().frame();
+    if (!frame)
+        return false;
```
I think it would be slightly better to only do this when we need frame, so when `isLink()` is true. That's also the only case we appear to be testing.

pratiksha-99 replied:
Makes sense. Updated.

rniwa commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,24 @@
+}
+
+function jsFuzzer() {
+  var dialogElement = document.getElementById("htmlvar00011");
```
Please use `const` or `let`.

rniwa commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,24 @@
+function jsFuzzer() {
+  var dialogElement = document.getElementById("htmlvar00011");
+
+  var newDocument = window.document.implementation.createHTMLDocument();
```
Ditto.

rniwa commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,24 @@
+</script>
+
+<body onload="jsFuzzer()">
+  <dialog id="htmlvar00011" ondragend="eventhandler5()">
```
Can this just be `div` instead of `dialog`?

pratiksha-99 replied:
`show()` is not a valid function for `div`. So keeping it `dialog` for the test.

rniwa commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,24 @@
+    testRunner.waitUntilDone();
+}
+
+function jsFuzzer() {
```
Maybe rename this to something more descriptive like `runTest`?

rniwa commented on LayoutTests/mathml/mathml-invalid-frame-crash.html:
```diff
@@ -0,0 +1,24 @@
+<p>This test passes if WebKit does not crash.</p>
```
It's odd to have this before body.
Please move to below `<body>`.

annevk commented on Source/WebCore/mathml/MathMLElement.cpp:
```diff
@@ -268,8 +268,12 @@ bool MathMLElement::isKeyboardFocusable(KeyboardEvent* event) const
+        RefPtr frame = document().frame();
+        if (!frame)
+            return false;
+        return frame->eventHandler().tabsToLinks(event);
```
Optional: this is a slightly nicer pattern to use here:
```suggestion
        return frame && frame->eventHandler().tabsToLinks(event);
```

pratiksha-99 committed:
Nullptr crash in WebCore::MathMLElement::isKeyboardFocusable
https://bugs.webkit.org/show_bug.cgi?id=288296
rdar://144405709

Reviewed by Anne van Kesteren and Ryosuke Niwa.

Added a nullptr check for the LocalFrame before focused keyboard events.

* LayoutTests/mathml/mathml-invalid-frame-crash-expected.txt: Added.
* LayoutTests/mathml/mathml-invalid-frame-crash.html: Added.
* Source/WebCore/mathml/MathMLElement.cpp:
(WebCore::MathMLElement::isKeyboardFocusable const):

Canonical link: https://commits.webkit.org/291060@main
